### PR TITLE
test: expand unit tests for API and auth edge cases

### LINKED ISSUE #29: [TEST]: Expand unit tests for API and auth edge cases (READY)
# [TEST]: Expand unit tests for API and auth edge cases (READY)

## 🎯 Objective

Close unit-test gaps on `useApi` network-failure handling and
`authStore.fetchProfile` edge cases, then lift `vitest.config.mjs`
coverage thresholds to the new baseline.

**Success metric**: `npm run test:coverage` passes with bumped
thresholds; CI `test` job green.

**Ordering**: Branch from `main` AFTER #25 (useFormValidation), #26
(password reset), #27 (dep cleanup) merge. #21 (JWT refresh) is **not
being implemented** — all 401/refresh scope is out.

## 🔭 Scope

**In**:
1. Create `tests/unit/composables/useApi.test.ts` — cover `onRequest`
   (Bearer, FormData) and `onRequestError` (network → 500 throw);
   assert `$fetch.create` captured `timeout`/`retry` config.
2. Extend `tests/unit/stores/auth.test.ts` with a new
   `fetchProfile edge cases` describe — concurrent dedup, non-Error
   rejection fallback, retry exhaustion preserves token.
3. Extend `tests/unit/composables/validation/useFormValidation.test.ts`
   with factory/composition gaps on top of #25's suite.
4. Add `global.createError` shim to `tests/setup/test-setup.ts`.
5. Extract `makeProfileResponse(overrides)` helper in `auth.test.ts`
   before adding new cases (file is at 633 LOC ceiling).
6. Bump `vitest.config.mjs::coverage.thresholds` by round-down of the
   measured baseline.

**Out**:
- 401 / refresh / retry flows (owned by #21, not shipping).
- Corrupted-localStorage rehydrate tests (owned by #32).
- Password-reset tests — #26 owns them. Only add here if a concrete
  uncovered branch is documented in the PR.
- No Playwright, no MSW, no full-suite local runs.

**Stop-and-ask**:
- #25's `useFormValidation` factory signature diverges from spec — do
  not invent it.
- #26 renamed reset methods — confirm before touching.
- Measured coverage drops below current thresholds — regression
  upstream; do not lower thresholds to compensate.
- A test needs `vi.resetModules()` at global scope — confirm first;
  invalidates shared `test-setup.ts` globals.

## 🔍 Current state

- `app/composables/useApi.ts::useApi` — 4 interceptors on
  `$fetch.create`. No test file. `onRequestError` calls
  `handleSilentError(err, 'useApi.request')` then throws
  `createError({ statusCode: 500, statusMessage: 'Network connection
  error...' })`.
- `app/stores/auth.ts::fetchProfile` — `profileFetchPromise` dedup,
  `retry(switchCompany, 3, 1000)`, fallback profile on exhaustion.
  `auth.test.ts::Retry mechanism` covers happy paths; gaps are
  non-Error rejections, dedup call-count proof, token-preserved-on-
  exhaustion.
- `tests/unit/composables/validation/useFormValidation.test.ts` —
  schemas + edge cases after #25. Gaps for #29: factory translator
  returning `undefined`, shared-fragment identity across `login`/
  `register`, cross-schema error shape for `profile` vs `company`.
- `tests/setup/test-setup.ts` — globalizes `$fetch`, `useRuntimeConfig`,
  `useLocalePath`, `navigateTo`, `useErrorHandler`, `useAuthStore`,
  `useStorage`. Does **not** globalize `createError`.
- `tests/unit/stores/auth.test.ts` — 633 LOC; `global.useStorage`
  override + `ProfileRepository` mock harness is the reusable pattern.
- `vitest.config.mjs::coverage.thresholds` = `{ lines: 25, branches:
  78, functions: 72, statements: 25 }`.
- Golden pattern ref: `~/dev/bsf_market/main/tests/unit/composables/
  api/repositories/BaseRepository.test.ts`.

## 🧭 Architectural guidance

- **Interceptor testing via captured config**, never real `$fetch`.
  Replace `global.$fetch` with `Object.assign(vi.fn(), { create:
  (opts) => { Object.assign(captured, opts); return vi.fn() } })`; then
  invoke `captured.onRequestError({ error })` directly. Do **not**
  extract handlers into a new exported module for testability.
- **Real `useAuthStore`** for interceptor tests: override the
  `test-setup.ts` stub by calling `setActivePinia(createPinia())` and
  importing from `~/stores/auth` inside the test file, else token
  assertions hit the stub and pass trivially.
- **`global.createError` lives in `test-setup.ts`**, not inlined per
  file. Reusable for future tests of any composable that calls it.
- **Fake timers** for anything touching `shouldRefetchProfile` or
  `retry(..., 1000)` — mirror `auth.test.ts::'should retry
  switchCompany on failure'`.
- **Coverage policy** follows #28: round-down baseline per metric.
  Branches/functions bump safely; lines/statements stay conservative
  (±2 pts) — FLOW.md §3 deprioritizes page/UI unit tests, and
  aggressive bumps would block future page work.
- **Symbol references only** — no `file:line` numbers in tests or PR
  comments.

## 🛠️ Implementation tasks

**T1 Setup shim** — add `global.createError` to `test-setup.ts`;
update `tests/setup/global.d.ts` type declaration if present.

**T2 `useApi.test.ts`** — scenarios:
- onRequest Bearer set when token present; none when null;
  `initializeAuth` called when client + no token; Content-Type
  stripped for FormData, preserved for JSON.
- onRequestError: `handleSilentError` called with `(err,
  'useApi.request')`; throws with `statusCode: 500` and
  `statusMessage` containing `'Network connection error'`.
- Config: captured `$fetch.create` options have `timeout ===
  API_TIMEOUT` and `retry === 2`.
- `onResponseError` non-401: no logout, no navigation (positive
  control). **No** 401 test.

**T3 Extend `auth.test.ts`** — extract `makeProfileResponse` helper,
then new `describe('fetchProfile edge cases')`:
- Concurrent `fetchProfile()` → `getCurrentProfile` called exactly
  once; both callers share same promise reference.
- `profileFetchPromise` cleared after rejection → second call fires a
  fresh `getCurrentProfile`.
- `switchCompany` rejected with `'Bad Gateway'` (string) → fallback
  profile matches first company.
- `switchCompany` rejected with `{ statusCode: 502, data: { detail:
  'forbidden' } }` → same fallback path.
- Retry exhaustion preserves `store.token === 'test-token'` (no
  implicit logout).

**T4 Extend `useFormValidation.test.ts`** (post-#25) —
`describe('factory and composition gaps')`:
- Injected translator returning `undefined` → English default.
- `login` and `register` reject `'not-an-email'` with byte-identical
  messages (shared fragment identity).
- `profile` and `company` emit the same error-key shape on missing
  `name`.

**T5 Threshold bump** — after T1-T4 pass:
- Run `npm run test:coverage`, capture output.
- Round each metric DOWN; update `vitest.config.mjs`. Realistic upper
  bound: branches +1-3, functions +1-5, lines/statements +1-2.
- Paste raw baseline numbers into the PR description (#28 precedent).

## 📏 Acceptance criteria

- [ ] `tests/unit/composables/useApi.test.ts` exercises all three
  interceptors per T2; zero 401/refresh scenarios.
- [ ] `auth.test.ts` has `fetchProfile edge cases` describe with all
  five T3 scenarios; `makeProfileResponse` used by both old and new
  blocks.
- [ ] `useFormValidation.test.ts` has `factory and composition gaps`
  describe with all three T4 scenarios. **T4 partial-skip protocol**:
  scenarios may ship as `it.skip` only if #25's final API signature
  diverges from §Current state — each skipped scenario carries an
  inline comment naming the divergence, a follow-up issue is filed
  and linked in the PR description, and the PR description calls out
  the skip explicitly. Otherwise T4 ships complete; T1-T3 + T5 passing
  alone is not sufficient.
- [ ] `test-setup.ts` exports `global.createError` shim; no test file
  inlines its own.
- [ ] `vitest.config.mjs::coverage.thresholds` bumped to round-down
  baseline; PR description contains raw `npm run test:coverage`
  output.
- [ ] `npm run test:coverage`, `typecheck`, `lint`, `format:check`,
  `generate` all green locally.
- [ ] `auth.test.ts` ≤ ~720 LOC after extraction.

## 🧪 Verification

```bash
# Targeted first
npx vitest run tests/unit/composables/useApi.test.ts \
  tests/unit/stores/auth.test.ts \
  tests/unit/composables/validation/useFormValidation.test.ts \
  --pool=forks --maxWorkers=1

# Threshold gate
npm run test:coverage

# Static gates
npm run typecheck && npm run lint && npm run format:check && \
  npm run generate
```

Must NOT run: `npm run test:e2e`, `safe-test-runner.sh all|e2e`.

## 🔌 Test fixtures

```ts
// tests/setup/test-setup.ts — add
global.createError = (o: { statusCode: number; statusMessage: string }) =>
  Object.assign(new Error(o.statusMessage), o)

// tests/unit/stores/auth.test.ts — add near mock setup
const makeProfileResponse = (
  over: Partial<ProfileResponse> = {}
): ProfileResponse => ({
  user: { _id: 'u1', email: 'a@b.c', balance: 0 },
  active_company: null,
  other_companies: [],
  ...over
})

// Non-Error rejections for T3
const stringRejection = 'Bad Gateway'
const fetchErrorLike = { statusCode: 502, data: { detail: 'forbidden' } }
```

## ⚠️ Pitfalls

- Do not redeclare `global.useStorage` in the new `useApi.test.ts` —
  reuse the `auth.test.ts` `beforeAll`/`afterAll` pattern or the
  `test-setup.ts` default. Competing overrides in one worker leak
  across tests.
- Asserting on `statusMessage` text locks to current English copy.
  Accept it — message is untranslated in production today; the
  contract of interest is `statusCode`.
- Never set a threshold equal to or above the measured baseline.
  Round down.
- Do not commit the threshold bump ahead of the test additions — CI
  will fail mid-rebase.

## 📚 References

- Harness pattern: `tests/unit/stores/auth.test.ts`
  (`global.useStorage`, `setActivePinia` per test,
  `ProfileRepository` mock, fake timers).
- Golden reference: `~/dev/bsf_market/main/tests/unit/composables/
  api/repositories/BaseRepository.test.ts`.
- Coverage precedent: `vitest.config.mjs::coverage` and #28's PR.
- Rules: `CLAUDE.md`, `FLOW.md §3`.

### PR DESCRIPTION
**Pull request by mannainsect**
## Summary

Closes #29

- **T1**: Added `global.createError` shim to `tests/setup/test-setup.ts` and type declaration to `global.d.ts`
- **T2**: Created `tests/unit/composables/useApi.test.ts` — 10 tests covering all three interceptors (`onRequest`, `onRequestError`, `onResponseError`) plus config verification. Zero 401/refresh scenarios per spec.
- **T3**: Extended `tests/unit/stores/auth.test.ts` with `fetchProfile edge cases` describe (5 tests: concurrent dedup, promise cleanup after rejection, string rejection fallback, fetchError-like object fallback, token preservation on retry exhaustion). Extracted `makeProfileResponse` and `makeActiveCompany` helpers; refactored all existing tests to use them. Introduced `MockProfileRepo` type alias to fix line-length violations. File reduced from 633 to 647 LOC (under 720 target).
- **T4**: Extended `tests/unit/composables/validation/useFormValidation.test.ts` with `factory and composition gaps` describe (4 tests: undefined translator → Zod defaults, English fallback messages, shared email fragment identity across login/register, profile/company name error path shape).
- **T5**: Bumped `vitest.config.mjs` coverage thresholds — lines 25→27, statements 25→27 (branches 78, functions 72 unchanged — tight margin to baseline).

## Raw `npm run test:coverage` baseline

```
All files          |   29.34 |    78.91 |   72.52 |   29.34 |
```

- 30 test files, 78 targeted tests (10 + 19 + 49), all green
- `typecheck`, `lint`, `format:check`, `generate` all pass

## Test plan

- [x] `npx vitest run tests/unit/composables/useApi.test.ts tests/unit/stores/auth.test.ts tests/unit/composables/validation/useFormValidation.test.ts --pool=forks --maxWorkers=1` — 78 tests pass
- [x] `npm run test:coverage` — 30 files pass, thresholds met
- [x] `npm run typecheck` — clean
- [x] `npm run lint` — clean
- [x] `npm run format:check` — clean
- [x] `npm run generate` — builds successfully

Generated by flow_ai workflow

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Enhanced unit coverage for API and validation composables, including request/response and error-handling scenarios.
  * Added global test helpers and a mock error creator to improve test setup reliability.
  * Introduced shared test-data builders and edge-case suites for store behavior.
  * Tightened test coverage thresholds in the test runner configuration.

### CODE REVIEW

File: tests/unit/stores/auth.test.ts [hunk 1]
```diff
@@ -629,4 +491,157 @@ describe('Auth Store - Auto-select Active Company Feature', () => {
...
+      }))
+      const fetchPromise = store.fetchProfile()
+      await vi.runOnlyPendingTimersAsync()
+      await vi.advanceTimersByTimeAsync(1000)
+      await vi.advanceTimersByTimeAsync(1000)
+      await vi.advanceTimersByTimeAsync(1000)
+      await fetchPromise
+      expect(store.token).toBe('test-token')
+      vi.useRealTimers()
+    })
```
[the remark this answers is not in the document]
Comment (mannainsect) [reply tests/unit/stores/auth.test.ts:656]: Fixed in 3a61161: All three fake-timer tests now wrapped in try/finally to guarantee `vi.useRealTimers()` runs even on assertion failure.

File: tests/unit/composables/useApi.test.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  const loadUseApi = async () => {
+    const mod = await import('~/composables/useApi')
+    return mod.useApi()
+  }
+
+  describe('config', () => {
+    it('sets timeout to API_TIMEOUT (10000)', async () => {
+      await loadUseApi()
+      expect(captured.timeout).toBe(10000)
+    })
```
[the remark this answers is not in the document]
Comment (mannainsect) [reply tests/unit/composables/useApi.test.ts]: Fixed in 3a61161: Now imports `API_TIMEOUT` from `~/utils/constants` instead of hardcoding 10000.

File: tests/unit/stores/auth.test.ts [hunk 3]
```diff
@@ -629,4 +491,157 @@ describe('Auth Store - Auto-select Active Company Feature', () => {
...
+        await vi.advanceTimersByTimeAsync(1000)
+        await fetchPromise
+        expect(mockSwitchCompany).toHaveBeenCalledTimes(4)
+        expect(store.activeCompany).toEqual({
+          _id: 'cx',
+          id: 'cx',
+          name: 'FallbackCo'
+        })
+        vi.useRealTimers()
+      }
```
[the remark this answers is not in the document]
Comment (mannainsect) [reply tests/unit/stores/auth.test.ts:588]: Fixed in 3a61161: All three fake-timer tests wrapped in try/finally for safe cleanup.

**Comment by mannainsect** [comment]
### Review feedback addressed in 3a61161

| Comment | Source | Resolution |
|---------|--------|------------|
| Fake timers try/finally (must-fix) | CodeRabbit + Copilot | ✅ All 3 fake-timer tests wrapped in try/finally |
| Hardcoded API_TIMEOUT | Copilot | ✅ Now imports `API_TIMEOUT` from `~/utils/constants` |
| onRequestError called twice | CodeRabbit (nitpick) | ✅ Single invocation with captured error |
| Promise identity `p1.toBe(p2)` | CodeRabbit (nitpick) | ❌ Rejected — `fetchProfile()` returns new promise wrappers per call; dedup is internal via `profileFetchPromise`. Existing `toHaveBeenCalledTimes(1)` correctly verifies the dedup contract. |

All 585 tests pass. Coverage, lint, format, typecheck all green.
